### PR TITLE
Reuse NDArray mem in TensorViews

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##
Reuse NDArray mem in TensorViews

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/storage/cpu_device_storage.h [hunk 1]
```diff
@@ -54,7 +54,11 @@ class CPUDeviceStorage {
   /*!
    * \brief Alignment of allocation.
    */
+#if MXNET_USE_NGRAPH == 1
+  static constexpr size_t alignment_ = 64;
```
Comment (mbrookhart) [inline src/storage/cpu_device_storage.h]: Is the alignment constant between ngraph backends?

File: src/ngraph/ngraph_nnvm_utils.h [hunk 2]
```diff
@@ -74,6 +74,24 @@ inline TensorViewVector make_ngraph_placeholders(
...
+    const std::vector<mxnet::OpReqType>* req = nullptr) {
+  TensorViewVector out;
+  for (size_t i = 0; i < ndarrays.size(); ++i) {
+    auto shape = TShape_to_NShape(ndarrays[i].shape());
+    const auto& element_type = getType(ndarrays[i].dtype());
+    if ((req != nullptr) && ((*req)[i] == mxnet::kAddTo))
+      out.push_back(backend->make_primary_tensor_view(element_type, shape));
+    else
+      out.push_back(backend->make_primary_tensor_view(
+          element_type, shape, ndarrays[i].storage_handle().dptr));
```
Comment (mbrookhart) [inline src/ngraph/ngraph_nnvm_utils.h]: Does this API work for Argon and GPU?

Comment (ashokei) [reply src/ngraph/ngraph_nnvm_utils.h]: ngraph plans to support this api for all backends, in one form or other. GPU could technically re-use CUDA mem , but NDArray integration may not be as straightforward.

Until we have them fully supported/stabilized, I will modify PR to do CPU backend for now.

Comment (ashokei) [reply src/ngraph/ngraph_nnvm_utils.h]: @csullivan can you please review these changes. MXNet has its own cuda mem management, and this PR should pass that cuda mem to ngraph. Is there a simple model for GPU test

Comment (csullivan) [reply src/ngraph/ngraph_nnvm_utils.h]: Yep, you can run examples/image_recognition/train_mnist.py when compiling with GPU support and setting MXNET_NGRAPH_GPU=TRUE

File: src/ngraph/ngraph_nnvm_utils.h [hunk 3]
```diff
@@ -74,6 +74,24 @@ inline TensorViewVector make_ngraph_placeholders(
   return out;
 }
 
+// creates and returns vector of TensorViews for corresponding NDArrays
+// reuses NDArray memory for each TensorView if req is not kAddTo
+inline TensorViewVector get_tensor_views(
```
[the remark this answers is not in the document]
Comment (ashokei) [reply src/ngraph/ngraph_nnvm_utils.h]: This specific function is similar to other inline functions in this header, but in larger context we could refactor out of header if these functions get out of hand.

[the remark this answers is not in the document]
Comment (mbrookhart) [reply src/ngraph/ngraph_nnvm_utils.h]: Originally, we put a lot of templates functions in the header files, and it ended up growing to more utility like functions. We could definitely refactor a lot of this into util.h/util.cc type files.
